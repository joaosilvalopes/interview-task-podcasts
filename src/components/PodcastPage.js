import { useEffect, useState } from "react";
import { useParams, Outlet } from "react-router-dom";
import styled from "styled-components";

import Card from './Card';

const Main = styled.main`
    display: flex;
    margin-top: 2rem;
`;

const PodcastCardContainer = styled.div`
    display: flex;
    justify-content: center;
    align-items: flex-start;
    width: 30%;
`;

const PodcastCard = styled(Card)`
    display: flex;
    flex-direction: column;
    padding: 1rem;
    width: 80%;
    max-width: 30rem;
`;

const PodcastImage = styled.img`
    width: 70%;
    align-self: center;
`;

const PodcastTitle = styled.p`
    font-weight: bold;
`;

const Author = styled.span`
    margin-top: 0.3rem;

    & > span {
        font-style: italic;
    }
`;

const PodcastDescriptionLabel = styled.p`
    font-weight: bold;
`;

const PodcastDescription = styled.p`
    margin-top: 0.75rem;
`;

const Separator = styled.hr`
    width: 100%;
    border: 0.05rem solid rgba(0, 0, 0, 0.2);
    margin: 1rem 0;
`;

const getCachedPodcast = (podcastId) => {
    try {
        const cachedPodcast = JSON.parse(localStorage.getItem('podcast' + podcastId));

        return cachedPodcast.ttl && Date.now() < cachedPodcast.ttl ? cachedPodcast.value : null;
    } catch {
        return null;
    }
}

const fetchPodcast = async (podcastId) => {
    const lookupRes = await fetch(`https://cors-anywhere.herokuapp.com/https://itunes.apple.com/lookup?id=${podcastId}&media=podcast&entity=podcastEpisode`).then(res => res.json());

    const [details, ...episodes] = lookupRes.results;

    const description = details.feedUrl ? await fetch(`https://cors-anywhere.herokuapp.com/${details.feedUrl}`)
        .then(response => response.text())
        .then(str => new window.DOMParser().parseFromString(str, "application/xml"))
        .then(xml => xml.querySelector('description').textContent) : '';

    const podcast = {
        title: details.collectionName,
        author: details.artistName,
        image: details.artworkUrl30,
        description,
        episodesById: episodes.reduce((acc, episode) => ({ ...acc, [episode.trackId]: { url: episode.episodeUrl, description: episode.description.replaceAll('\n', '<br />'), title: episode.trackName, length: episode.trackTimeMillis, releaseDate: episode.releaseDate } }), {}),
        episodeCount: episodes.length
    };

    localStorage.setItem('podcast' + podcastId, JSON.stringify({ value: podcast, ttl: Date.now() + 86000000 }));

    return podcast;
}

const PodcastPage = () => {
    const { podcastId } = useParams();
    const [podcast, setPodcast] = useState(getCachedPodcast(podcastId));

    useEffect(() => {
        if (podcast) {
            return;
        }

        fetchPodcast(podcastId).then(setPodcast).catch(console.log);
    }, [podcastId, podcast, setPodcast]);

    return podcast ? (
        <Main>
            <PodcastCardContainer>
                <PodcastCard>
                    <PodcastImage src={podcast.image} />
                    <Separator />
                    <PodcastTitle>{podcast.title}</PodcastTitle>
                    <Author>by <span>{podcast.author}</span></Author>
                    <Separator />
                    <PodcastDescriptionLabel>Description:</PodcastDescriptionLabel>
                    <PodcastDescription style={{ whiteSpace: 'pre-wrap' }}>{podcast.description}</PodcastDescription>
                </PodcastCard>
            </PodcastCardContainer>
            <Outlet context={{ podcast }} />
        </Main>
    ) : null;
}

export default PodcastPage;