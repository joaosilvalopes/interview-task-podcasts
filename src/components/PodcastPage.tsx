import { useContext, useEffect, useState } from "react";
import { useParams, Outlet } from "react-router-dom";
import styled from "styled-components";

import LoadingContext from "../context/LoadingContext";
import fetchWithCors from "../utils/fetchWithCors";

import Card from './Card';

const Main = styled.main`
    display: flex;
    padding: 3.2rem;
    font-size: 1.6rem;
`;

const PodcastCardContainer = styled.div`
    display: flex;
    justify-content: flex-start;
    align-items: flex-start;
    width: 30%;
`;

const PodcastCard = styled(Card)`
    display: flex;
    flex-direction: column;
    padding: 1.6rem;
    width: 80%;
    max-width: 50rem;
`;

const PodcastImage = styled.img`
    width: 70%;
    align-self: center;
    margin: 1rem 0;
`;

const PodcastTitle = styled.p`
    font-weight: bold;
`;

const Author = styled.span`
    margin-top: 0.5rem;

    & > span {
        font-style: italic;
    }
`;

const PodcastDescriptionLabel = styled.p`
    font-weight: bold;
`;

const PodcastDescription = styled.p`
    margin-top: 1.2rem;
`;

const Separator = styled.hr`
    width: 100%;
    border: 0.1rem solid rgba(0, 0, 0, 0.1);
    margin: 1.6rem 0;
`;

type PodcastType = {
    title: string,
    author: string,
    image: string,
    description: string,
    episodesById: {
        [key: string]: {
            title: string,
            description: string,
            url: string,
            releaseDate: string,
            length: number,
        }
    },
    episodeCount: number
};

type PodcastResType = { results: [
    {
        collectionName: string,
        artistName: string,
        artworkUrl100: string,
        artworkUrl600: string,
        description: string,
        length: number,
        feedUrl: string,
    },
    ...{
        trackId: string,
        episodeUrl: string,
        description: string,
        trackName: string,
        trackTimeMillis: number,
        releaseDate: string,
    }[],
] };

const getCachedPodcast = (podcastId: string): PodcastType|null => {
    try {
        const storedPodcastJson = localStorage.getItem('podcast' + podcastId);
        const cachedPodcast = storedPodcastJson && JSON.parse(storedPodcastJson);

        return cachedPodcast.ttl && Date.now() < cachedPodcast.ttl ? cachedPodcast.value : null;
    } catch {
        return null;
    }
}

const fetchPodcast = async (podcastId: string) => {
    const lookupRes: PodcastResType = await fetchWithCors(`https://itunes.apple.com/lookup?id=${podcastId}&media=podcast&entity=podcastEpisode`).then(res => res.json());

    const [details, ...episodes] = lookupRes.results;

    const description = details.feedUrl ? await fetchWithCors(details.feedUrl)
        .then(response => response.text())
        .then(str => new window.DOMParser().parseFromString(str, "application/xml"))
        .then(xml => xml.querySelector('description')?.textContent || '') : '';

    const podcast = {
        title: details.collectionName,
        author: details.artistName,
        image: `${details.artworkUrl100} 768w, ${details.artworkUrl600} 1200w`,
        description: description.replaceAll('\n', '<br />'),
        episodesById: episodes.reduce((acc, episode) => ({ ...acc, [episode.trackId]: { url: episode.episodeUrl, description: episode.description.replaceAll('\n', '<br />'), title: episode.trackName, length: episode.trackTimeMillis, releaseDate: episode.releaseDate } }), {}),
        episodeCount: episodes.length
    };

    localStorage.setItem('podcast' + podcastId, JSON.stringify({ value: podcast, ttl: Date.now() + 86000000 }));

    return podcast;
}

const PodcastPage = () => {
    const { podcastId } = useParams();
    const [podcast, setPodcast] = useState(getCachedPodcast(podcastId as string));
    const [, setLoading] = useContext(LoadingContext);

    useEffect(() => {
        if (podcast) {
            setLoading(false);

            return;
        }

        setLoading(true);

        fetchPodcast(podcastId as string).then(setPodcast).catch(console.log).finally(() => setLoading(false));
    }, [podcastId, podcast, setPodcast, setLoading]);

    return podcast ? (
        <Main>
            <PodcastCardContainer>
                <PodcastCard data-testid="podcast">
                    <PodcastImage data-testid="podcast-image" srcSet={podcast.image} />
                    <Separator />
                    <PodcastTitle data-testid="podcast-title">{podcast.title}</PodcastTitle>
                    <Author>by <span data-testid="podcast-author">{podcast.author}</span></Author>
                    <Separator />
                    <PodcastDescriptionLabel>Description:</PodcastDescriptionLabel>
                    <PodcastDescription  data-testid="podcast-description" dangerouslySetInnerHTML={{ __html: podcast.description }} />
                </PodcastCard>
            </PodcastCardContainer>
            <Outlet context={{ podcast }} />
        </Main>
    ) : null;
}

export default PodcastPage;