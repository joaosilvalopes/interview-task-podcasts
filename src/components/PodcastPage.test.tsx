import { render, waitFor } from '@testing-library/react';
import { RouterProvider, createMemoryRouter } from "react-router-dom";

import PodcastPage from './PodcastPage';

it('should render correctly', async () => {
    const router = createMemoryRouter([{
        path: "/podcast/:podcastId",
        element: <PodcastPage />,
    }],  { initialEntries: [ '/podcast/1664273280' ] });
    const { asFragment, getByTestId } = render(<RouterProvider router={router} />);

    await waitFor(() =>expect(getByTestId('podcast')), { timeout: 10000 });

    expect(getByTestId('podcast-title')).toBeVisible();
    expect(getByTestId('podcast-author')).toBeVisible();
    expect(getByTestId('podcast-description')).toBeVisible();
    
    expect(asFragment()).toMatchSnapshot();
});
