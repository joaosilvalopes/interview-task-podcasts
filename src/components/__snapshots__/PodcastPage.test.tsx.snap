// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<DocumentFragment>
  <main
    class="sc-gswNZR cSdtLP"
  >
    <div
      class="sc-dkrFOg bxwRPC"
    >
      <div
        class="sc-bcXHqe sc-hLBbgP bFLobp hucncV"
        data-testid="podcast"
      >
        <img
          class="sc-eDvSVe bgaNpC"
          srcset="https://is1-ssl.mzstatic.com/image/thumb/Podcasts126/v4/29/cc/b0/29ccb00c-5970-2efb-18e9-1ea340ee1f9d/mza_17732988801572498978.jpg/100x100bb.jpg 768w, https://is1-ssl.mzstatic.com/image/thumb/Podcasts126/v4/29/cc/b0/29ccb00c-5970-2efb-18e9-1ea340ee1f9d/mza_17732988801572498978.jpg/600x600bb.jpg 1200w"
        />
        <hr
          class="sc-pyfCe bWpAUh"
        />
        <p
          class="sc-jSUZER dkeOOJ"
          data-testid="podcast-title"
        >
          The Coldest Case In Laramie
        </p>
        <span
          class="sc-gKPRtg btLwBr"
        >
          by 
          <span
            data-testid="podcast-author"
          >
            Serial Productions & The New York Times
          </span>
        </span>
        <hr
          class="sc-pyfCe bWpAUh"
        />
        <p
          class="sc-iBYQkv hzcOrO"
        >
          Description:
        </p>
        <p
          class="sc-ftTHYK bqlHXM"
          data-testid="podcast-description"
        >
          Kim Barker, a Pulitzer Prize-winning investigative reporter for The New York Times, revisits an unsolved murder that took place while she was in high school in Laramie, Wyoming, nearly 40 years ago. She confronts the conflicting stories people have told themselves about the crime because of an unexpected development: the arrest of a former Laramie police officer accused in the murder.
        </p>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
